### PR TITLE
[21649] Correct iterator increment after erasing elements in `connected_servers_list`

### PR DESCRIPTION
**Pull request by LuciaEchevarria99**
## Description

This PR addresses a critical issue with iterator handling in the loop that processes elements of `connected_servers_`. The original implementation encountered an assertion failure due to invalid iterator increments after removing elements from the container.

Comment for the reviewer:
- Alternative approach to fix the problem: Given that the `connected_servers_` list is expected to contain at most one element with the matching `guidPrefix`, another viable solution could be to use a break statement within the `if (it->guidPrefix == pdata->m_guid.guidPrefix)` condition. This would immediately exit the loop after finding and removing the element, eliminating the need to handle iterator invalidation explicitly.

@Mergifyio backport 2.14.x 2.10.x

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- __NO__: Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
    - The team decided internally to add regression test in a follow-up PR
- N/A: Any new/modified methods have been properly documented using Doxygen. 
- N/A: Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- [x] Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- [x] Changes are API compatible. 
- N/A: New feature has been added to the `versions.md` file (if applicable).
- N/A: New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [x] Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- N/A: If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 1]
```diff
@@ -594,7 +594,7 @@ void PDPClient::removeRemoteEndpoints(
         eprosima::shared_lock<eprosima::shared_mutex> disc_lock(mp_builtin->getDiscoveryMutex());
 
         // Verify if this participant is a server
-        for (auto it = connected_servers_.begin(); it != connected_servers_.end(); ++it)
+        for (auto it = connected_servers_.begin(); it != connected_servers_.end();)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp:597]:
```suggestion
        auto it = connected_servers_.begin();
        while (it != connected_servers_.end())
```

File: src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp [hunk 2]
```diff
@@ -603,6 +603,10 @@ void PDPClient::removeRemoteEndpoints(
                 is_server = true;
                 mp_sync->restart_timer(); // enable announcement and sync mechanism till this server reappears
             }
+            else
+            {
+                ++it;
+            }
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPClient.cpp:609]:
```suggestion
                mp_sync->restart_timer(); // enable announcement and sync mechanism till this server reappears
                
                // Avoid incrementing iterator after item removal
                continue;
            }
            
            ++it;
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Woohh! Nice catch!

Is it possible to have a regression test?

**Review by MiguelCompany — APPROVED** [review]

**Comment by Mario-DL** [comment]
Mac failed test is a known issue, unrelated to this PR

**Comment by JesusPoderoso** [comment]
@Mergifyio backport 2.14.x 2.10.x

**Pull request merged**